### DISCUSSION TITLE
Making from github clone causes "The site cannot provide a secure connection" and SSL cipher mismatch

### DISCUSSION THREAD

**Discussion by RussellHamker** [cockpit-project/cockpit#17925] [General]
I am working on setting up cockpit on a VMware Photon OS 4 system. After cloning the repo and building, installing, and then running cockpit, I get an error when attempting to access the site from the local network from another PC using Chrome and Firefox browser. I have tried via FQDN and IP with no luck.

the text of the error message (or a screenshot of it)
See below

which version of Linux you're installing Cockpit on
Photon OS 4

your browser (and browser version)
Chrome 104.0.5304.107 (windows client)
Firefox 106.0.5 (windows client)

Are you using a self-signed certificate (default) or a custom certificate?
using the cert that was generated when the make/make install was completed

Are you using a proxy?
No

Are you connecting from a different machine from the server?
Yes

Are you connecting through the local network or over the Internet?
Local network

![image](https://user-images.githubusercontent.com/102260516/202221815-76077450-7963-4ea0-a2e5-54c63ad5707b.png)

**Comment by RussellHamker** [comment]
Here are the install steps I am taking:

#### Prep for cockpit

tdnf install -y json-glib json-glib-devel gnutls-devel xmlto krb5 krb5-devel polkit polkit-devel libssh libssh-devel Linux-PAM Linux-PAM-devel build-essential git libxml2 libxml2-devel e2fsprogs-devel nodejs

#### Clone cockpit
git clone https://github.com/cockpit-project/cockpit.git

cd cockpit/

#### Build cockpit
./autogen.sh --prefix=/usr --enable-debug --disable-pcp

#### Make cockpit
make #completes with errors make -t fixes
make -t
make

#### Install cockpit
make install

#### Open firewall port
sudo iptables -I INPUT -p tcp -m tcp --dport 9090 -j ACCEPT
sudo iptables-save > /etc/systemd/scripts/ip4save

#### Enable Cockpit socket
sudo systemctl enable --now cockpit.socket

**Comment by RussellHamker** [comment]
When looking at the service logs, I see the below error:

Nov 16 08:42:54 photontkg.guarddog.lab cockpit-tls[8848]: cockpit-tls: gnutls_handshake failed: A packet with illegal or unsupported version was received.
Nov 16 08:42:54 photontkg.guarddog.lab cockpit-tls[8848]: cockpit-tls: gnutls_handshake failed: A packet with illegal or unsupported version was received.
Nov 16 08:42:54 photontkg.guarddog.lab cockpit-tls[8848]: cockpit-tls: gnutls_handshake failed: A packet with illegal or unsupported version was received.
Nov 16 08:42:54 photontkg.guarddog.lab cockpit-tls[8848]: cockpit-tls: gnutls_handshake failed: A packet with illegal or unsupported version was received.
Nov 16 08:44:24 photontkg.guarddog.lab systemd[1]: cockpit.service: Succeeded.
░░ Subject: Unit succeeded
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ The unit cockpit.service has successfully entered the 'dead' state.

**Comment by RussellHamker** [comment]
Errors showing in journal

journalctl -xeu cockpit

-- Journal begins at Mon 2022-11-14 12:11:04 MST, ends at Wed 2022-11-16 09:39:01 MST. --
Nov 16 09:39:01 photontkg.guarddog.lab systemd[1]: Starting Cockpit Web Service...
░░ Subject: A start job for unit cockpit.service has begun execution
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ A start job for unit cockpit.service has begun execution.
░░
░░ The job identifier is 826.
Nov 16 09:39:01 photontkg.guarddog.lab cockpit-certificate-ensure[4193]: /usr/libexec/cockpit-certificate-helper: line 25: sscg: command not found
Nov 16 09:39:01 photontkg.guarddog.lab cockpit-certificate-ensure[4194]: .........+...+.+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*......+...+...+..+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++>
Nov 16 09:39:01 photontkg.guarddog.lab cockpit-certificate-ensure[4194]: ...+.....+....+...........+.........+.........+......+....+...+........+....+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*...+........+.......+>
Nov 16 09:39:01 photontkg.guarddog.lab cockpit-certificate-ensure[4194]: -----
Nov 16 09:39:01 photontkg.guarddog.lab systemd[1]: Started Cockpit Web Service.
░░ Subject: A start job for unit cockpit.service has finished successfully
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ A start job for unit cockpit.service has finished successfully.
░░
░░ The job identifier is 826.
Nov 16 09:39:01 photontkg.guarddog.lab cockpit-tls[4200]: cockpit-tls: gnutls_handshake failed: A packet with illegal or unsupported version was received.
Nov 16 09:39:01 photontkg.guarddog.lab cockpit-tls[4200]: cockpit-tls: gnutls_handshake failed: A packet with illegal or unsupported version was received.
Nov 16 09:39:01 photontkg.guarddog.lab cockpit-tls[4200]: cockpit-tls: gnutls_handshake failed: A packet with illegal or unsupported version was received.
Nov 16 09:39:01 photontkg.guarddog.lab cockpit-tls[4200]: cockpit-tls: gnutls_handshake failed: A packet with illegal or unsupported version was received.
lines 1-25/25 (END)

**Comment by RussellHamker** [comment]
I have found if I update the gnutls default TLS versions to be different, I can get the web interface to operate.

#### Set TLS Defaults to 1.2 and use secure256 (disable tls 1.1)
echo 'SYSTEM=NONE:!VERS-SSL3.0:!VERS-TLS1.0:-VERS-TLS-ALL:!VERS-TLS1.1:+VERS-TLS1.2:+VERS-TLS1.3:+SECURE256:+AES-128-CCM-8:+AES-128-CBC:+RSA:+SHA1:+COMP-NULL' > /etc/gnutls/default-priorities

**Comment by RussellHamker** [comment]
Now for the better question on this. Why does the web interface NOT default to localhost when you don't have anything filled in? It should always default to localhost so that users dont have to type in a system name to connect locally.

**Reply by garrett** [reply]
I use `localhost:9090` every day. Sometimes even `127.0.0.1:9090`. Other times, my machine's LAN hostname (for when I check on a phone). It should work for you, but you were having other strange issues.

**Comment by RussellHamker** [comment]
The Storage area does not work:
![image](https://user-images.githubusercontent.com/102260516/202471242-2f5d7ddc-0871-490c-aed5-cb728637012e.png)

**Reply by garrett** [reply]
The Storage page needs `storaged` installed. I don't see it available in Photon OS 4 when I run `tdnf search storaged` in my Photon OS VM.

**Reply by jelly** [reply]
@RussellHamker compiled their own Cockpit, which tries to load `kdump.js` which does not seem compatible with what we expect and errors out.

**Comment by garrett** [comment]
Cockpit doesn't officially support every distribution. It's up to your distribution to port it, if we're not actively working on it.

We have a list of "supported" distributions that we specifically do things to ensure Cockpit works on a distribution:

https://cockpit-project.org/running.html

The list is:

- Fedora
- Red Hat Enterprise Linux
- Fedora CoreOS
- CentOS
- Debian
- Ubuntu

Additionally, these distributions do their own testing and work with us to submit patches, but we don't actively test them in Cockpit's testing framework. (These distributions do their own testing, as do the others above as well.):

- Clear Linux
- Arch Linux
- openSUSE Tumbleweed

Several distributions and various custom spins of the above are based on one of the above and inherit Cockpit working for their distribution. (Such as Alma Linux, Rocky Linux, Linux Mint, Fedora Silverblue, Fedora IoT, etc.)

However, VMware's Photon OS is a new distribution not made from any of the above. That said, it has a C re-implementation of `dnf` (as `tdnf`, as you know) and uses RPM, like Fedora, RHEL, CentOS Stream, and even openSUSE (which has different packages).

This morning, I looked if it would be "easy" to take the CentOS Stream 9 src.rpm and rebuild on Photon OS. I tried this out in a VM. I was hoping there was a way to grab build-time packages to compile a src.rpm, like there is on Fedora, RHEL, CentOS, etc. (`sudo dnf builddep thePackageName.src.rpm` and then `rpmbuild rebuild thePackageName.src.rpm`). Ideally, you'd eventually want to have Cockpit packaged in RPMs for easier deployment, of course. But getting it to build at all (like you're doing) is a good first step, of course.

I didn't see anything like builddep. There was an `rpm-build` package to install, though. It would require making sure all the dependencies are met, and just because a distribution is based on RPM, doesn't mean it's fully compatible. (Famously, SUSE distributions use different naming conventions and different packages than Red Hat based distributions, even though they're both based on RPM. However, as Photon OS has `tdnf` based on `dnf`, I would assume it's much more Fedora and Red Hat like in naming.)

While searching, I did find [an article for Photon OS 3 mentioning that Cockpit is in photo-extras](https://jreypo.io/2016/01/12/you-can-run-cockpit-in-photon-os/). However, it's a very old article (2016) with an old version of Photon OS and old version of Cockpit too, and it looks like that's not the case for Photon OS 4 and recent versions of Cockpit. But that might be a start to get things working again?

Overall, the issues you're experiencing are ones I have not seen on other distributions (and including what people have reported, not my own personal experiences). There's probably something strange in a lower level library or perhaps Cockpit might even be making some assumptions about how things work (based on the supported distributions) that could be different on Photo OS.

It looks like you're trying to get Cockpit up and running on Photon OS 4 and that's a great effort. We would welcome patches and feedback on how to get things working on Photon OS. 

We do have documentation on how to develop for Cockpit @:
https://cockpit-project.org/external/wiki/Contributing.html

Perhaps this specific document about compiling Cockpit might help? https://cockpit-project.org/external/source/HACKING.html

If you do figure out the underlying issue or anything tricky, it would be great to hear more about that. Thanks!

We do hang out on IRC at `#cockpit` on `irc.libera.chat`. It might be useful to ask some questions there in a more "realtime" environment. (We're generally there on weekdays during European business hours.) More info about using IRC is @ https://libera.chat/ and you can connect via that website if you don't already have a client. I think the `#cockpit` channel requires you to have a registered "nick", however (but that's relatively easy once you're on IRC: https://libera.chat/guides/registration). 

I hope this helps! And best of luck getting Cockpit up and running on Photon OS 4.
